File: lib/services/api_server.py
# filename: api_server.py
from fastapi import FastAPI
from pydantic import BaseModel
from joblib import load

app = FastAPI()
model = load("knn_model.joblib")

class SensorData(BaseModel):
    soil_moisture: float
    temperature: float
    soil_humidity: float
    ph: float
    rainfall: float
    air_humidity: float

@app.post("/predict")
def predict(data: SensorData):
    features = [[
        data.soil_moisture,
        data.temperature,
        data.soil_humidity,
        data.ph,
        data.rainfall,
        data.air_humidity
    ]]
    prediction = model.predict(features)[0]
    return {"needs_watering": prediction}

File: lib/services/api_server.dart
import 'package:http/http.dart' as http;
import 'dart:convert';

class ApiService {
  static Future<bool?> getWateringRecommendation({
    required double soilMoisture,
    required double temperature,
    required double soilHumidity,
    required double ph,
    required double rainfall,
    required double airHumidity,
  }) async {
    final url = Uri.parse('http://192.168.1.10:8000/predict');

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          "soil_moisture": soilMoisture,
          "temperature": temperature,
          "soil_humidity": soilHumidity,
          "ph": ph,
          "rainfall": rainfall,
          "air_humidity": airHumidity,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['needs_water'] == 1;
      } else {
        print("Prediction failed: ${response.statusCode}");
        return null;
      }
    } catch (e) {
      print("Error calling prediction API: $e");
      return null;
    }
  }
}
